1. Scenario: A company wants to analyze the sales performance of its products in different regions. They have collected the following data:

   Region A: [10, 15, 12, 8, 14]
   
   Region B: [18, 20, 16, 22, 25]
   
   Calculate the mean sales for each region.


In [1]:
region_a_sales = [10, 15, 12, 8, 14]
region_b_sales = [18, 20, 16, 22, 25]

mean_a = sum(region_a_sales) / len(region_a_sales)
mean_b = sum(region_b_sales) / len(region_b_sales)

print("Mean sales for Region A:", mean_a)
print("Mean sales for Region B:", mean_b)


Mean sales for Region A: 11.8
Mean sales for Region B: 20.2


2. Scenario: A survey is conducted to measure customer satisfaction on a scale of 1 to 5. The data collected is as follows:

   [4, 5, 2, 3, 5, 4, 3, 2, 4, 5]
   
   Calculate the mode of the survey responses.


In [2]:
from collections import Counter

# Survey responses
responses = [4, 5, 2, 3, 5, 4, 3, 2, 4, 5]

# Count the frequency of each value
frequency = Counter(responses)

# Find the maximum frequency
max_frequency = max(frequency.values())

# Find the mode(s) with the maximum frequency
mode = [value for value, freq in frequency.items() if freq == max_frequency]

print("Mode(s) of the survey responses:", mode)


Mode(s) of the survey responses: [4, 5]


3. Scenario: A company wants to compare the salaries of two departments. The salary data for Department A and Department B are as follows:
   Department A: [5000, 6000, 5500, 7000]
   
   Department B: [4500, 5500, 5800, 6000, 5200]
   
   Calculate the median salary for each department.


To calculate the median salary for each department, we need to find the middle value(s) of the sorted salary data. If the number of values is odd, the median is the middle value. If the number of values is even, the median is the average of the two middle values. Let's calculate it step by step for each department:

For Department A: [5000, 6000, 5500, 7000]
1. Sort the salaries in ascending order: [5000, 5500, 6000, 7000]
2. The number of values is odd (4), so the median is the middle value.
   Median of Department A: 5500

For Department B: [4500, 5500, 5800, 6000, 5200]
1. Sort the salaries in ascending order: [4500, 5200, 5500, 5800, 6000]
2. The number of values is odd (5), so the median is the middle value.
   Median of Department B: 5500

Therefore, the median salary for Department A is 5500, and the median salary for Department B is also 5500.

4. Scenario: A data analyst wants to determine the variability in the daily stock prices of a company. The data collected is as follows:

   [25.5, 24.8, 26.1, 25.3, 24.9]
   
   Calculate the range of the stock prices.


To calculate the range of the stock prices, we need to find the difference between the maximum and minimum values in the dataset. In this case, the dataset is:

[25.5, 24.8, 26.1, 25.3, 24.9]

Let's calculate it step by step:

1. Sort the stock prices in ascending order: [24.8, 24.9, 25.3, 25.5, 26.1]
2. The minimum value is the first value in the sorted list: 24.8
3. The maximum value is the last value in the sorted list: 26.1
4. Calculate the range by subtracting the minimum value from the maximum value: 26.1 - 24.8 = 1.3

Therefore, the range of the stock prices is 1.3.

5. Scenario: A study is conducted to compare the performance of two different teaching methods. The test scores of the students in each group are as follows:
   Group A: [85, 90, 92, 88, 91]
   
   Group B: [82, 88, 90, 86, 87]
   Perform a t-test to determine if there is a significant difference in the mean scores between the two groups.


In [4]:
from scipy import stats

# Test scores for Group A and Group B
group_a_scores = [85, 90, 92, 88, 91]
group_b_scores = [82, 88, 90, 86, 87]

# Perform independent t-test
t_statistic, p_value = stats.ttest_ind(group_a_scores, group_b_scores)

# Print the t-statistic and p-value
print("T-statistic:", t_statistic)
print("P-value:", p_value)


T-statistic: 1.4312528946642733
P-value: 0.19023970239078333


The t-statistic represents the magnitude of the difference between the means of the two groups, while the p-value indicates the probability of observing such a difference by chance. In this case, with a p-value of 0.190, if we consider a significance level of 0.05 (commonly used), the p-value is greater than the significance level. Therefore, we do not have sufficient evidence to reject the null hypothesis and can conclude that there is no significant difference in the mean scores between the two groups.

6. Scenario: A company wants to analyze the relationship between advertising expenditure and sales. The data collected is as follows:
   Advertising Expenditure (in thousands): [10, 15, 12, 8, 14]
   
   Sales (in thousands): [25, 30, 28, 20, 26]
   
   Calculate the correlation coefficient between advertising expenditure and sales.


In [5]:
import numpy as np

# Advertising expenditure (in thousands) and sales (in thousands)
advertising_expenditure = [10, 15, 12, 8, 14]
sales = [25, 30, 28, 20, 26]

# Calculate the correlation coefficient
correlation_coefficient = np.corrcoef(advertising_expenditure, sales)[0, 1]

# Print the correlation coefficient
print("Correlation coefficient:", correlation_coefficient)


Correlation coefficient: 0.8757511375750132


The correlation coefficient ranges from -1 to 1. A positive value indicates a positive correlation, meaning that as advertising expenditure increases, sales tend to increase as well. In this case, the correlation coefficient of  0.8757 suggests a strong positive correlation between advertising expenditure and sales.

7. Scenario: A survey is conducted to measure the heights of a group of people. The data collected is as follows:

   [160, 170, 165, 155, 175, 180, 170]
   
   Calculate the standard deviation of the heights.


In [6]:
import numpy as np

# Heights data
heights = [160, 170, 165, 155, 175, 180, 170]

# Calculate the standard deviation
standard_deviation = np.std(heights)

# Print the standard deviation
print("Standard deviation of heights:", standard_deviation)


Standard deviation of heights: 7.953949089757174


8. Scenario: A company wants to analyze the relationship between employee tenure and job satisfaction. The data collected is as follows:
   Employee Tenure (in years): [2, 3, 5, 4, 6, 2, 4]
   
   Job Satisfaction (on a scale of 1 to 10): [7, 8, 6, 9, 5, 7, 6]
   
   Perform a linear regression analysis to predict job satisfaction based on employee tenure.


In [7]:
from sklearn.linear_model import LinearRegression
import numpy as np

# Employee tenure (in years) and job satisfaction
employee_tenure = np.array([2, 3, 5, 4, 6, 2, 4]).reshape(-1, 1)
job_satisfaction = np.array([7, 8, 6, 9, 5, 7, 6])

# Create a linear regression model
regression_model = LinearRegression()

# Fit the model to the data
regression_model.fit(employee_tenure, job_satisfaction)

# Get the coefficients of the linear regression equation
slope = regression_model.coef_[0]
intercept = regression_model.intercept_

# Print the linear regression equation
print("Linear regression equation: job_satisfaction = {} * employee_tenure + {}".format(slope, intercept))


Linear regression equation: job_satisfaction = -0.4680851063829787 * employee_tenure + 8.595744680851062


The linear regression equation represents the relationship between employee tenure and job satisfaction. In this case, the negative coefficient (-0.468) indicates that as employee tenure increases, job satisfaction tends to decrease slightly. The intercept term (8.5957) represents the estimated job satisfaction when the employee tenure is zero.

You can use this equation to predict job satisfaction for new values of employee tenure.

9. Scenario: A study is conducted to compare the effectiveness of two different medications. The recovery times of the patients in each group are as follows:
   Medication A: [10, 12, 14, 11, 13]
   
   Medication B: [15, 17, 16, 14, 18]
   
   Perform an analysis of variance (ANOVA) to determine if there is a significant difference in the mean recovery times between the two medications.


In [8]:
from scipy import stats

# Recovery times for Medication A and Medication B
medication_a_times = [10, 12, 14, 11, 13]
medication_b_times = [15, 17, 16, 14, 18]

# Perform one-way ANOVA
f_statistic, p_value = stats.f_oneway(medication_a_times, medication_b_times)

# Print the F-statistic and p-value
print("F-statistic:", f_statistic)
print("P-value:", p_value)


F-statistic: 16.0
P-value: 0.003949772803445326


The F-statistic tests the null hypothesis that the means of the two groups are equal. The p-value indicates the probability of observing such a difference in means by chance. In this case, with a p-value of 0.003, if we consider a significance level of 0.05 (commonly used), the p-value is less than the significance level. Therefore, we have marginal evidence to reject the null hypothesis and conclude that there may be a significant difference in the mean recovery times between the two medications. However, the decision ultimately depends on the predetermined significance level and other factors such as sample size and study design.

10. Scenario: A company wants to analyze customer feedback ratings on a scale of 1 to 10. The data collected is

 as follows:
    [8, 9, 7, 6, 8, 10, 9, 8, 7, 8]
    
    Calculate the 75th percentile of the feedback ratings.


In [1]:
import numpy as np

ratings = [8, 9, 7, 6, 8, 10, 9, 8, 7, 8]

percentile_75 = np.percentile(ratings, 75)
print("75th percentile: ", percentile_75)


75th percentile:  8.75


11. Scenario: A quality control department wants to test the weight consistency of a product. The weights of a sample of products are as follows:

    [10.2, 9.8, 10.0, 10.5, 10.3, 10.1]
    
    Perform a hypothesis test to determine if the mean weight differs significantly from 10 grams.


In [2]:
from scipy.stats import ttest_1samp

weights = [10.2, 9.8, 10.0, 10.5, 10.3, 10.1]
population_mean = 10

t_statistic, p_value = ttest_1samp(weights, population_mean)

alpha = 0.05  # significance level

print("t-statistic:", t_statistic)
print("p-value:", p_value)

if p_value < alpha:
    print("Reject null hypothesis. The mean weight differs significantly from 10 grams.")
else:
    print("Fail to reject null hypothesis. The mean weight does not differ significantly from 10 grams.")


t-statistic: 1.5126584522688367
p-value: 0.19077595151110102
Fail to reject null hypothesis. The mean weight does not differ significantly from 10 grams.


12. Scenario: A company wants to analyze the click-through rates of two different website designs. The number of clicks for each design is as follows:
    Design A: [100, 120, 110, 90, 95]
    
    Design B: [80, 85, 90, 95, 100]
    
    Perform a chi-square test to determine if there is a significant difference in the click-through rates between the two designs.


In [3]:
from scipy.stats import chi2_contingency

design_a = [100, 120, 110, 90, 95]
design_b = [80, 85, 90, 95, 100]

observed = [design_a, design_b]

chi2, p_value, dof, expected = chi2_contingency(observed)

alpha = 0.05  # significance level

print("Chi-square statistic:", chi2)
print("p-value:", p_value)

if p_value < alpha:
    print("Reject null hypothesis. There is a significant difference in click-through rates between the two designs.")
else:
    print("Fail to reject null hypothesis. There is no significant difference in click-through rates between the two designs.")


Chi-square statistic: 6.110658166925435
p-value: 0.19103526314060296
Fail to reject null hypothesis. There is no significant difference in click-through rates between the two designs.


13. Scenario: A survey is conducted to measure customer satisfaction with a product on a scale of 1 to 10. The data collected is as follows:

    [7, 9, 6, 8, 10, 7, 8, 9, 7, 8]
    
    Calculate the 95% confidence interval for the population mean satisfaction score.


In [4]:
import numpy as np
from scipy.stats import t

data = [7, 9, 6, 8, 10, 7, 8, 9, 7, 8]
confidence_level = 0.95

sample_mean = np.mean(data)
sample_std = np.std(data, ddof=1)  # ddof=1 for sample standard deviation
sample_size = len(data)

critical_value = t.ppf((1 + confidence_level) / 2, df=sample_size - 1)
standard_error = sample_std / np.sqrt(sample_size)

confidence_interval = sample_mean + critical_value * standard_error

lower_bound = sample_mean - critical_value * standard_error
upper_bound = sample_mean + critical_value * standard_error

print("95% Confidence Interval: [{:.2f}, {:.2f}]".format(lower_bound, upper_bound))


95% Confidence Interval: [7.04, 8.76]


14. Scenario: A company wants to analyze the effect of temperature on product performance. The data collected is as follows:
    Temperature (in degrees Celsius): [20, 22, 23, 19, 21]
    
    Performance (on a scale of 1 to 10): [8, 7, 9, 6, 8]
    
    Perform a simple linear regression to predict performance based on temperature.


In [5]:
import numpy as np
from scipy import stats

temperature = np.array([20, 22, 23, 19, 21])
performance = np.array([8, 7, 9, 6, 8])

slope, intercept, r_value, p_value, std_err = stats.linregress(temperature, performance)

print("Slope:", slope)
print("Intercept:", intercept)
print("R-squared:", r_value**2)
print("p-value:", p_value)
print("Standard error:", std_err)


Slope: 0.5
Intercept: -2.9000000000000004
R-squared: 0.4807692307692307
p-value: 0.19417134561205843
Standard error: 0.30000000000000004


15. Scenario: A study is conducted to compare the preferences of two groups of participants. The preferences are measured on a Likert scale from 1 to 5. The data collected is as follows:
    Group A: [4, 3, 5, 2, 4]
    
    Group B: [3, 2, 4, 3, 3]
    
    Perform a Mann-Whitney U test to determine if there is a significant difference in the median preferences between the two groups.



In [6]:
from scipy.stats import mannwhitneyu

group_a = [4, 3, 5, 2, 4]
group_b = [3, 2, 4, 3, 3]

statistic, p_value = mannwhitneyu(group_a, group_b, alternative='two-sided')

alpha = 0.05  # significance level

print("Mann-Whitney U statistic:", statistic)
print("p-value:", p_value)

if p_value < alpha:
    print("Reject null hypothesis. There is a significant difference in median preferences between the two groups.")
else:
    print("Fail to reject null hypothesis. There is no significant difference in median preferences between the two groups.")


Mann-Whitney U statistic: 17.0
p-value: 0.380836480306712
Fail to reject null hypothesis. There is no significant difference in median preferences between the two groups.


16. Scenario: A company wants to analyze the distribution of customer ages. The data collected is as follows:
    [25, 30, 35, 40, 45, 50, 55, 60, 65, 70]
    
    Calculate the interquartile range (IQR) of the ages.


In [7]:
import numpy as np

ages = [25, 30, 35, 40, 45, 50, 55, 60, 65, 70]

Q1 = np.percentile(ages, 25)
Q3 = np.percentile(ages, 75)

IQR = Q3 - Q1

print("Interquartile Range (IQR):", IQR)


Interquartile Range (IQR): 22.5


17. Scenario: A study is conducted to compare the performance of three different machine learning algorithms. The accuracy scores for each algorithm are as follows:

    Algorithm A: [0.85, 0.80, 0.82, 0.87, 0.83]
    
    Algorithm B: [0.78, 0.82, 0.84, 0.80, 0.79]
    
    Algorithm C: [0.90, 0.88, 0.89, 0.86, 0.87]
    Perform a Kruskal-Wallis test to determine if there is a significant difference in the median accuracy scores between the algorithms.


In [8]:
from scipy.stats import kruskal

algorithm_a = [0.85, 0.80, 0.82, 0.87, 0.83]
algorithm_b = [0.78, 0.82, 0.84, 0.80, 0.79]
algorithm_c = [0.90, 0.88, 0.89, 0.86, 0.87]

statistic, p_value = kruskal(algorithm_a, algorithm_b, algorithm_c)

alpha = 0.05  # significance level

print("Kruskal-Wallis statistic:", statistic)
print("p-value:", p_value)

if p_value < alpha:
    print("Reject null hypothesis. There is a significant difference in median accuracy scores between the algorithms.")
else:
    print("Fail to reject null hypothesis. There is no significant difference in median accuracy scores between the algorithms.")


Kruskal-Wallis statistic: 9.696947935368053
p-value: 0.007840333026249539
Reject null hypothesis. There is a significant difference in median accuracy scores between the algorithms.


18. Scenario: A company wants to analyze the effect of price on sales. The data collected is as follows:
    Price (in dollars): [10, 15, 12, 8, 14]
    
    Sales: [100, 80, 90, 110, 95]
    Perform a simple linear regression to predict

 sales based on price.


In [9]:
import numpy as np
from scipy import stats

price = np.array([10, 15, 12, 8, 14])
sales = np.array([100, 80, 90, 110, 95])

slope, intercept, r_value, p_value, std_err = stats.linregress(price, sales)

print("Slope:", slope)
print("Intercept:", intercept)
print("R-squared:", r_value**2)
print("p-value:", p_value)
print("Standard error:", std_err)


Slope: -3.506097560975609
Intercept: 136.3719512195122
R-squared: 0.8064024390243901
p-value: 0.03850178234753776
Standard error: 0.9918303504036147



19. Scenario: A survey is conducted to measure the satisfaction levels of customers with a new product. The data collected is as follows:

    [7, 8, 9, 6, 8, 7, 9, 7, 8, 7]
    
    Calculate the standard error of the mean satisfaction score


To calculate the standard error of the mean satisfaction score, you need to divide the standard deviation of the data by the square root of the sample size. Here are the steps:

1. Calculate the mean of the satisfaction scores:
   Mean = (7 + 8 + 9 + 6 + 8 + 7 + 9 + 7 + 8 + 7) / 10 = 76 / 10 = 7.6

2. Calculate the differences between each satisfaction score and the mean:
   Differences = [7 - 7.6, 8 - 7.6, 9 - 7.6, 6 - 7.6, 8 - 7.6, 7 - 7.6, 9 - 7.6, 7 - 7.6, 8 - 7.6, 7 - 7.6]
               = [-0.6, 0.4, 1.4, -1.6, 0.4, -0.6, 1.4, -0.6, 0.4, -0.6]

3. Square each difference:
   Squared Differences = [(-0.6)^2, 0.4^2, 1.4^2, (-1.6)^2, 0.4^2, (-0.6)^2, 1.4^2, (-0.6)^2, 0.4^2, (-0.6)^2]
                      = [0.36, 0.16, 1.96, 2.56, 0.16, 0.36, 1.96, 0.36, 0.16, 0.36]

4. Calculate the variance by taking the average of the squared differences:
   Variance = (0.36 + 0.16 + 1.96 + 2.56 + 0.16 + 0.36 + 1.96 + 0.36 + 0.16 + 0.36) / 10
            = 9.08 / 10
            = 0.908

5. Calculate the standard deviation by taking the square root of the variance:
   Standard Deviation = √0.908 ≈ 0.953

6. Calculate the standard error of the mean by dividing the standard deviation by the square root of the sample size:
   Standard Error = 0.953 / √10 ≈ 0.301

Therefore, the standard error of the mean satisfaction score is approximately 0.301.

20. Scenario: A company wants to analyze the relationship between advertising expenditure and sales. The data collected is as follows:
    Advertising Expenditure (in thousands): [10, 15, 12, 8, 14]
    Sales (in thousands): [25, 30, 28, 20, 26]
    Perform a multiple regression analysis to predict sales based on advertising expenditure.



In [12]:
! pip install statsmodels

import numpy as np
import statsmodels.api as sm

advertising_expenditure = np.array([10, 15, 12, 8, 14])
sales = np.array([25, 30, 28, 20, 26])

# Add a constant term to the predictor variable
X = sm.add_constant(advertising_expenditure)

# Create the OLS (Ordinary Least Squares) model
model = sm.OLS(sales, X)

# Fit the model to the data
results = model.fit()

# Print the summary of the regression analysis
print(results.summary())


     ---------------------------------------- 0.0/9.2 MB ? eta -:--:--
     ---------------------------------------- 0.1/9.2 MB 6.4 MB/s eta 0:00:02
      --------------------------------------- 0.1/9.2 MB 1.4 MB/s eta 0:00:07
      --------------------------------------- 0.2/9.2 MB 1.3 MB/s eta 0:00:07
      --------------------------------------- 0.2/9.2 MB 1.2 MB/s eta 0:00:08
     - -------------------------------------- 0.3/9.2 MB 1.0 MB/s eta 0:00:09
     - -------------------------------------- 0.3/9.2 MB 1.1 MB/s eta 0:00:09
     - -------------------------------------- 0.3/9.2 MB 1.0 MB/s eta 0:00:09
     - -------------------------------------- 0.4/9.2 MB 1.0 MB/s eta 0:00:09
     - -------------------------------------- 0.4/9.2 MB 981.2 kB/s eta 0:00:09
     - -------------------------------------- 0.5/9.2 MB 938.8 kB/s eta 0:00:10
     -- ------------------------------------- 0.5/9.2 MB 932.7 kB/s eta 0:00:10
     -- ------------------------------------- 0.5/9.2 MB 945.8 kB

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.767
Model:                            OLS   Adj. R-squared:                  0.689
Method:                 Least Squares   F-statistic:                     9.872
Date:                Thu, 13 Jul 2023   Prob (F-statistic):             0.0516
Time:                        16:49:29   Log-Likelihood:                -9.5288
No. Observations:                   5   AIC:                             23.06
Df Residuals:                       3   BIC:                             22.28
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         12.2012      4.429      2.755      0.0

C:\Users\viksb\miniconda3\lib\site-packages\statsmodels\stats\stattools.py:74: ValueWarning: omni_normtest is not valid with less than 8 observations; 5 samples were given.
  warn("omni_normtest is not valid with less than 8 observations; %i "
